### PR TITLE
In place formulas

### PR DESCRIPTION
**Pull request by naterush**
# Description

This PR begins to add support for in-place formula editing. The specification is here: https://www.notion.so/trymito/In-place-formula-writing-b1eed93706fe4a1e8112b95039d26f8b

The main development is that formulas now properly overwrite the previous if these formulas are back to back. Note that this does not happen with step id, but rather through the more complex step skipping mechanisms in the backend. Notably, I had to fix some bugs in this mechanism to get it working.

Overall, I think this will be a pretty usability solution for most users in practice. If we explicitly usability test too unrealistically, I think we'll notice confusion that never will appear in practice -- so let's make sure we design the usability test to be realistic!

# Testing

Write some new formulas. Test them with undo, redo, etc.

# Documentation

We need to write new formula docs explaining our differences! If we're happy with this, I'll write them.

### CODE REVIEW

**Comment by aarondr77** [comment]
Here are a few initial thoughts:

# Reapplying a formula + ephemeral formulas 

## Problem Analysis

Consider the scenario where you have the formula in column , =A+1. Currently, if you open the formula and then resubmit it, it has the effect of applying = A+ 2 to the original dataset. That is pretty confusing and not expected. **If you don't change the formula of a self-referential formula, you don't expect the value to change.** 

In another scenario, if you set the formula of column B to =A+1, then you only expect that result to change if the values in column A has updated, and that is the case! If the value of A has not changed, then no matter how many times you resubmit this formula, the result will always be the same.

## Problem formalization

Resubmitting formulas and having the value change when the formula didn't change is only a concern for certain types of formulas. For it to be the case, the formula must be:
1. Self referential ie: Resubmitting the formula for A as =B+1 a bunch of times won't change the value of A. 
2. A non-terminal formula (this isn't the right term). 

Consider the following formulas for column A:

#### Formulas that return the same values across resubmits:
-- `UPPER(A)`
-- `LEFT(A, 5)`
-- `TEXT(A)`

#### Formulas that won't return the same values across resubmits:
-- `LEFT(A, FIND(A, '-') - 2)` -> the found value will go from some number to 0
-- `IF(A > B, 'greater', 'lesser')` -> This will error, comparing a string to a number
-- `A + 1`
-- `MONTH(A)` -> This will error because A is now an int instead of a datetime. 
-- `AVG(A)`

Based on [this quick analysis](https://www.notion.so/trymito/In-place-formula-writing-b1eed93706fe4a1e8112b95039d26f8b#3df0259b95ed4439854c110853d74024), about 57% of formulas that users submit fall into the "Formulas that won't return the same values across resubmits".

## Proposals 

#### Proposal 1
I think a reasonable, although it comes with its own set of UX problems, UX would be that if you apply a self referential formula, that formula doesn't persist after its application. This is ideal for the situation where you set the formula for column A to =A + 1. Then, the only way to increment the values again is to explicitly recreate the formula. 

The 2 big UX issues here are:
1. Just around re-introducing two types of formulas / columns back into Mito. It might be confusing that some formulas disappear after submitting and others don't. 
2. More technical implementation work.
3. Formulas that go away make it hard to go trace what has happened. 

#### Iteration ideation
1. Instead of just having the formula disappear completely, show the last submitted value in the cell editor too, and mark it as a previously submitted formula. ie: Don't put it in the formula bar. 
2. Have all formulas disapear after they were submitted. And again, make it easy to go back to the previous formula. This would also take care of "Set B to equal A. Delete A. The formula now makes no sense." & "Set B equal to A. Then change the dtype of B." This takes the "we don't have to worry about the order of steps" even further.  But it would make it difficult to refresh formulas, so we might want a refresh button or something. 
3. Have the formula persist + use the same stepID until another step has been made. So, if the user write =A+1 submits, and reopoens it and submits, it still has the effect of just =A+1 because it was the same stepID. Once the user makes an additional edit, the next time they open the cell editor for A, it is just =A or '' or something. This allows users to iterate on their formulas while they are writing them, but not get confused later by resubmitting the formula and having the value change. EDIT: The stepID persisting is already in this implementation :)

**Review by naterush — COMMENTED** [review]
Cleanup.

**Comment by aarondr77** [comment]
The X button in the tour doesn't close the tour if you click it on the first or second tour step, it just takes you to the last one. We had this previously because we wanted to take you to the Tutorial slide no matter what at the end. But it doesn't make sense in this context.

**Comment by aarondr77** [comment]
Ohhh I just had some strange behavior where I wrote some formulas than pressed the Undo button, then tried to reopen the cell editor by pressing Enter right away, but it instead applied the Undo again.

**Comment by aarondr77** [comment]
Can we give a better error message when the user tries to submit a formula that relies on a column that doesn't exist. If the user creates a new column and tries to reference a column that doesn't exist we give a specific error message, can we display that same error message if the user is resubmitting a previous formula too? 

Screen Shot 2022-06-09 at 10 13 47 AM

**Comment by aarondr77** [comment]
I opened the cell editor on a column that previously didn't have a formula applied, and then closed the cell editor by pressing enter again. It cleared all of the data in my column, which was super not expected. If I didn't edit the column, I didn't expect it to edit my data. 

This was additionally confusing because the cell editor looked empty, but the formula submitted was =0. 

If the user submit ='' as the formula, can we do nothing?

File: mitosheet/mitosheet/steps_manager.py [hunk 1]
```diff
@@ -60,7 +60,9 @@ def execute_step_list_from_index(
 
     If start_index is not given, will start from the initialize step.
     """
-    if start_index is None or start_index < 0:
```
[the remark this answers is not in the document]
Comment (naterush) [reply mitosheet/mitosheet/steps_manager.py:63]: Yep!

**Pull request merged**